savant/utils/memory_repr.py

```diff
@@ -0,0 +1,368 @@
+            )
+
+        if device is None:
+            device = 'cuda'
+
+        if device == 'cuda':
+            if img.ndim == 2 or input_format == 'channels_last':
+                return cupy_as_opencv_gpu_mat(img)
+
+            cuda_interface = img.__cuda_array_interface__
+            mat_type = _numpy_type_to_opencv_mat_type(
+                numpy_type=img.__cuda_array_interface__['typestr'],
+                channels=img.__cuda_array_interface__['shape'][channel_index],
+            )
+            if img.strides[channel_index] != int(
+                img.__cuda_array_interface__['typestr'][-1]
+            ):
+                if channel_index == 0:
+                    img = img.transpose(1, 2, 0)
+                img = img.ravel()
+
+            return cv2.cuda.createGpuMatFromCudaMemory(
+                (
+                    cuda_interface['shape'][width_index],
+                    cuda_interface['shape'][height_index],
+                ),
+                mat_type,
+                img.__cuda_array_interface__['data'][0],
+            )
+
+        elif device == 'cpu':
+            if channel_index == 2:
+                return img.get()
+            return cp.transpose(img, (1, 2, 0)).get()
+        else:
+            raise TypeError(
+                f'Unsupported device {device} to convert into OpenCV GPU image.'
+            )
+
+    raise TypeError(f'Unsupported type {type(img)} to convert into OpenCV GPU image.')
+
+
+def as_cupy(
+    img,
+    input_format: Optional[str] = None,
+    output_format: Optional[str] = None,
+) -> cp.ndarray:
+    """Converts PyTorch or OpenCv image to CuPy GPU image.
+
+    :param img: is a image in PyTorch or OpenCV format.
+    :param input_format: is a shape format of the input image
+        (`channels_first` or `channels_last`). `channels_last` is used as default
+        for cupy arrays. Parameter is ignored for OpenCV images.
+    :param output_format: is a shape format of the output image
+        (`channels_first` or `channels_last`). If output_format is None the format is
+        the same as format of the input image.
+    """
+    # numpy
+    if isinstance(img, np.ndarray):
+        cupy_image = cp.asarray(img)
+
+        if img.ndim == 2:
+            return cupy_image
+
+        if (
+            input_format == 'channels_last' or input_format is None
+        ) and output_format == 'channels_first':
+            cupy_image = cp.transpose(cupy_image, (2, 0, 1))
+        elif input_format == 'channels_first' and output_format == 'channels_last':
+            cupy_image = cp.transpose(cupy_image, (1, 2, 0))
+        return cupy_image
+
+    # opencv
+    elif isinstance(img, cv2.cuda.GpuMat):
+        cupy_image = opencv_gpu_mat_as_cupy(img)
+
+        if img.channels == 1:
+            return cupy_image
+
+        if output_format == 'channels_first':
+            cupy_image = cp.transpose(cupy_image, (2, 0, 1))
+        return cupy_image
+
+    # torch
+    try:
```

Comment:
Consider moving the "import" to the branch where it is needed and reduce the number of "if"-s. A simple refactoring of "as_cupy" (done in the last commit) gives a 15x speedup.

```
np_arr = np.random.rand(720, 1280).astype(dtype=np.float32)
cv_arr = cv2.cuda_GpuMat()
cv_arr.upload(np_arr)

%%timeit
cp_arr1 = as_cupy_old(cv_arr)
# 60.3 µs ± 533 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

%%timeit
cp_arr1 = as_cupy_new(cv_arr)
# 4.09 µs ± 26.3 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
```

Reply:
@dorgun @denisvmedyantsev I believe importing inside the function body is not a good idea and must be avoided. It is better to move functions from a flat space to subpackages involving certain imports on initialization.

Reply:
Redesigned to use specialized functions. Functions for working with pytorch are placed in a separate file.